lib/linked_list.rb

```diff
@@ -12,120 +12,372 @@ def initialize(value, next_node = nil)
 
 # Defines the singly linked list
 class LinkedList
-    def initialize
-      @head = nil # keep the head private. Not accessible outside this class
+  def initialize
+    @head = nil # keep the head private. Not accessible outside this class
+  end
+
+  # method to add a new node with the specific data value in the linked list
+  # insert the new node at the beginning of the linked list
+  def add_first(value)
+    current = @head
+    # add new node with provided value and set next to current
+    new_node = Node.new(value, current)
+    # new node becomes head
+    @head = new_node
+  end
+
+  # returns the value in the first node
+  # returns nil if the list is empty
+  def get_first
+    # if the list is empty, head is nil
+    if @head.nil?
+      return nil
+    else 
+      value = @head.data
+      return value
+    end
+  end
+
+  # method to find if the linked list contains a node with specified value
+  # returns true if found, false otherwise
+  def search(value)
+    current = @head
+
+    until current.nil?  
+      return true if current.data == value
+      current = current.next
+    end
+
+    return false
+    
+  end
+
+  # method to return the max value in the linked list
+  # returns the data value and not the node
+  def find_max
+    
+    if @head.nil?
+      return nil 
+    end
+
+    current = @head
+    max = current.data
+
+    until current.nil?
+      if current.data > max 
+        max = current.data
+      end
+      current = current.next
+    end
+
+    return max
+  end
+
+  # method to return the min value in the linked list
+  # returns the data value and not the node
+  def find_min
+
+    if @head.nil?
+      return nil 
+    end
+
+    # current = @head
+    min = @head.data
+    current = @head
+
+    # was failing tests until I realized stop condition for min/max traverse needs to be AFTER the end of list (current.nil?) not last node in list (current.next.nil?)
+    until current.nil?
+      if current.data < min 
+        min = current.data
+      end
+      current = current.next
+    end
+
+    return min
+  end
+
+  # method that returns the length of the singly linked list
+  def length
+    if @head.nil?
+      return 0
+    else
+      length = 1
+      current = @head
+      until current.next.nil?
+        current = current.next
+        length += 1
+      end
+    return length
     end
+  end
+
+  # method that returns the value at a given index in the linked list
+  # index count starts at 0
+  # returns nil if there are fewer nodes in the linked list than the index value
+  def get_at_index(index)
 
-    # method to add a new node with the specific data value in the linked list
-    # insert the new node at the beginning of the linked list
-    def add_first(value)
-      raise NotImplementedError
+    # determine length
+    if @head.nil?
+      return nil
+    else
+      length = 1
+      current = @head
+      until current.next.nil?

```

Comment:
I just want to note that this requires you to traverse the list twice, maybe you can avoid that.  Also if you're finding the length of the list, you can reuse the `.length` method.